### ISSUE TITLE
[BUG] tests fail due to TF32

### ISSUE THREAD

**Issue by davidkoski** [ml-explore/mlx-swift-lm#357]
On devices with a NAX (e.g. M5) these tests fail:

```swift
    @Test(
        "Chunked and single-pass prefill agree on logits and cache offsets",
        arguments: [3, 5, 8, 16])
    func chunkSizeInvariance(chunkSize: Int) throws {
```

```swift
    @Test("Gemma4 Unified text-only prefill is chunk-size invariant")
    func textOnlyPrefillChunkSizeInvariant() throws {
```

Repro:

- fails when run on M5
- works when `MLX_ENABLE_TF32=0` is set

**Comment by aleroot** [comment]
This looks like the same TF32/NAX test-oracle issue we already guarded against in SpeculativeDecodingTests. These Gemma4 tests build synthetic models with default float32 MLXNN Linear/Embedding weights, then compare chunked prefill vs single-pass prefill. Those paths are mathematically equivalent, but on NAX TF32 can choose different matmul/attention kernels/reduction order by shape, so the logits can drift past the tight tolerance.

I don’t think this points to a Gemma4 cache/prepare bug. More likely the test should cast the synthetic model weights to f16/bf16, or otherwise avoid strict float32 chunk-equivalence when TF32 is enabled.

I do not have an M5, so I cannot test, I am waiting for the MacStudio M5 😀

**Comment by davidkoski** [comment]
Agreed, that was my read as well.

**Comment by GoodOlClint** [comment]
Reproduced on an M5 Max and confirmed the TF32 diagnosis: running Gemma4ChunkedPrefillTests standalone via xcodebuild, the suite failed 5/5 runs with TF32 on (3–4 of the 4 chunk sizes over tolerance each run, always the logits `close` expectation — the cache-offset assertions never fail) and passed 5/5 with `MLX_ENABLE_TF32=0`. The per-run variation in which chunk sizes fail comes from the random weight init, which is why it looks flaky in full-suite runs.

I tried the fix suggested above — casting the synthetic model to f16 the way #326 did for SpeculativeDecodingTests — and it isn't sufficient on its own: at the existing `rtol: 1e-4, atol: 1e-5` the tests then fail in *both* TF32 modes. Once the model computes in f16, chunked and single-pass prefill run different-shaped matmuls whose kernels accumulate in different orders, and that noise alone exceeds the f32-calibrated tolerance. I instrumented the test to print max abs/rel logits drift across 16 runs × 4 chunk sizes: max absolute drift ~2e-2, in both TF32 modes (max relative is meaningless here — near-zero logits inflate it).

What worked: f16 cast + widening the comparison to `rtol: 1e-2, atol: 5e-2` (2.5× margin over the worst observed drift). With that, 10/10 runs pass with TF32 on and 10/10 with it off. The oracle still discriminates: the bugs this test exists to catch (cache/mask/chunk-boundary errors) perturb the logits at O(1), well past 5e-2.

Happy to open a PR with the two-part change if that approach is agreeable — mainly wanted input on the tolerance loosening first, since 5e-2 atol on logits is a bigger number than these tests have used and there may be a preference for a different oracle (e.g. comparing in a higher-precision dtype or on top-k agreement) instead.

**Comment by aleroot** [comment]
This confirms the issue is the test oracle, not the cache/chunking logic.

The f16 cast makes sense and matches #326. I would also make the model initialization deterministic, otherwise the measured drift and failing chunk sizes will keep changing between runs.

Since relative error is noisy around zero, I would prefer a direct max absolute diff check over `allClose`. If the deterministic worst case is still around `2e-2`, then `5e-2` seems reasonable with the exact cache-offset assertions retained.

Please go ahead with the PR.
